tests/test_base_device.py

```diff
@@ -0,0 +1,947 @@
+
+    def test_execute_pipeline_no_primitives_mocked(self, mocker):
+        """Test that a device **not** using Primitives only calls the _execute_runtime_service
+        to execute, regardless of measurement type"""
+
+        dev = QiskitDevice2(
+            wires=5, backend=backend, use_primitives=False, session=MockSession(backend)
+        )
+
+        initial_session = dev._session
+
+        sampler_execute = mocker.spy(dev, "_execute_sampler")
+        estimator_execute = mocker.spy(dev, "_execute_estimator")
+
+        qs = QuantumScript(
+            [qml.PauliX(0), qml.PauliY(1)],
+            measurements=[
+                qml.expval(qml.PauliZ(0)),
+                qml.probs(wires=[0, 1]),
+                qml.counts(),
+                qml.sample(),
+            ],
+        )
+
+        with patch.object(dev, "_execute_runtime_service", return_value="runtime_execute_res"):
+            runtime_service_execute = mocker.spy(dev, "_execute_runtime_service")
+            res = dev.execute(qs)
+
+        runtime_service_execute.assert_called_once()
+        sampler_execute.assert_not_called()
+        estimator_execute.assert_not_called()
+
+        assert res == "runtime_execute_res"
+        assert initial_session == dev._session  # session is not changed
+
+    @patch("pennylane_qiskit.qiskit_device2.QiskitDevice2._execute_estimator")
+    def test_execute_pipeline_primitives_no_session(self, mocker):
+        """Test that a Primitives-based device initialized with no Session creates one for the
+        execution, and then returns the device session to None."""
+
+        dev = QiskitDevice2(wires=5, backend=backend, use_primitives=True, session=None)
+
+        assert dev._session is None
+
+        qs = QuantumScript([qml.PauliX(0), qml.PauliY(1)], measurements=[qml.expval(qml.PauliZ(0))])
+
+        with patch("pennylane_qiskit.qiskit_device2.Session") as mock_session:
+            res = dev.execute(qs)
+            mock_session.assert_called_once()  # a session was created
+
+        assert dev._session is None  # the device session is still None
+
+    def test_execute_pipeline_with_all_execute_types_mocked(self, mocker):
+        """Test that a device that **is** using Primitives calls the _execute_runtime_service
+        to execute measurements that require raw samples, and the relevant primitive measurements
+        on the other measurements"""
+
+        dev = QiskitDevice2(
+            wires=5, backend=backend, use_primitives=True, session=MockSession(backend)
+        )
+
+        qs = QuantumScript(
+            [qml.PauliX(0), qml.PauliY(1)],
+            measurements=[
+                qml.expval(qml.PauliZ(0)),
+                qml.probs(wires=[0, 1]),
+                qml.counts(),
+                qml.sample(),
+            ],
+        )
+        tapes, reorder_fn = split_measurement_types(qs)
+
+        with patch.object(dev, "_execute_runtime_service", return_value="runtime_execute_res"):
```

Comment:
I think you can get this `as obj_name` to avoid the `mocker.spy` below and keep it simple, but idk :p